library/src/lib/date-picker/date-picker.component.ts

```diff
@@ -165,64 +161,87 @@ export class DatePickerComponent implements OnInit, OnDestroy, ControlValueAcces
     }
 
     /** Closes the calendar if it is open */
-    closeCalendar() {
+    public closeCalendar() {
         if (this.isOpen) {
             this.isOpen = false;
         }
     }
 
-    /** @hidden */
-    updateDatePickerInputHandler(d) {
-        if (this.type === 'single') {
-            if (d.selectedDay.date) {
-                const newInputDate = this.dateAdapter.format(d.selectedDay.date);
-                if (this.inputFieldDate !== newInputDate) {
-                    this.inputFieldDate = newInputDate;
-                    this.selectedDay = d.selectedDay;
-                    this.selectedDayChange.emit(this.selectedDay);
-                    this.onChange({date: this.selectedDay.date});
-                }
+    public handleSingleDateChange(date: FdDate) {
+        if (date) {
+            const newInputDate = this.dateAdapter.format(date);
+            if (this.inputFieldDate !== newInputDate) {
+                this.inputFieldDate = newInputDate;
             }
-        } else {
-            if (d.selectedFirstDay.date) {
-                const newInputDates = this.dateAdapter.format(d.selectedFirstDay.date) + this.dateAdapter.rangeDelimiter
-                    + this.dateAdapter.format(d.selectedLastDay.date);
-                if (this.inputFieldDate !== newInputDates) {
-                    this.inputFieldDate = newInputDates;
-                    this.selectedRangeFirst = d.selectedFirstDay;
-                    this.selectedRangeLast = d.selectedLastDay;
-                    this.selectedRangeFirstChange.emit(this.selectedRangeFirst);
-                    this.selectedRangeLastChange.emit(this.selectedRangeLast);
-                    this.onChange({date: this.selectedRangeFirst.date, rangeEnd: this.selectedRangeLast.date});
-                }
+            this.selectedDay = date;
+            this.selectedDayChange.emit(date);
+            // Of Course it will be changed, but without it, it throws: ExpressionChangedAfterItHasBeenCheckedError
+            setTimeout(() => {
+                this.onChange({ date: date });
+            }, 0);
+        }
+    }
+
+    public handleRangeDateChange(dates: { start: FdDate, end: FdDate }) {
+        if (dates) {
+            const newInputDates = this.dateAdapter.format(dates.start) + this.dateAdapter.rangeDelimiter
+                + this.dateAdapter.format(dates.end)
+            ;
+            if (this.inputFieldDate !== newInputDates) {
+                this.inputFieldDate = newInputDates;
             }
+            this.selectedRangeFirst = dates.start;
+            this.selectedRangeLast = dates.end;
+            this.selectedRangeFirstChange.emit(dates.start);
+            this.selectedRangeLastChange.emit(dates.end);
+            // Of Course it will be changed, but without it, it throws: ExpressionChangedAfterItHasBeenCheckedError
+            setTimeout(() => {
+                this.onChange({ date: dates.start, rangeEnd: dates.end });
```

Comment:
Might be because the event this is called on is emitted during an OnChange cycle